keps/20200901-standard-crd.md

```diff
@@ -0,0 +1,240 @@
+
+## Motivation
+
+There are a lot of different projects in the wild that essentially do the same thing: sync secrets with Kubernetes. The idea is to unify efforts into a single project that serves the needs of all users in this problem space.
+
+As a starting point i would like to define a **common denominator** for a CustomResourceDefinition that serves all known use-cases. This CRD should follow the standard alpha -> beta -> GA feature process.
+
+Once the CRD API is defined we can move on with more delicate discussions about technology, organization and processes.
+
+### Goals
+
+- Define a alpha CRD
+- Fully document the Spec and use-cases
+
+### Non-Goals
+
+This KEP proposes the CRD Spec and documents the use-cases, not the choice of technology or migration path towards implementing the CRD.
+
+## Terminology
+
+* Kubernetes External Secrets `KES`: A Application that runs a control loop which syncs secrets
+* KES `instance`: A single entity that runs a control loop.
+* ExternalSecret `ES`: A CustomResource that declares which secrets should be synced
+* Store: Is a **source** for secrets. The Store is external to KES. It can be a hosted service like Alibabacloud SecretsManager, AWS SystemsManager, Azure KeyVault...
+* Frontend: A **sink** for the synced secrets. Usually a `ConfigMap` or `Secret`
+* Secret: credentials that act as a key to sensitive information
+
+## Use-Cases
+* one global KES instance that manages ES in **all namespaces**, which gives access to **all stores**, with ACL
+* multiple global KES instances, each manages access to a single or multiple stores (e.g.: shard by stage or team...)
+* one KES per namespace (a user manages his/her own KES instance)
+
+### User definitions
+* `operator :=` i manage one or multiple `KES` instances
+* `user :=` i only create `ES`, KES is managed by someone else
+
+### User Stories
+From that we can derive the following requirements or user-stories:
+1. AS a KES operator i want to run multiple KES instances per cluster (e.g. one KES instance per DEV/PROD)
+2. AS a KES operator or user i want to integrate **multiple stores** from a **single KES instance** (e.g. dev namespace has access only to dev secrets)
+3. AS a KES user i want to control the sink for the secrets (aka frontend: store secret as `kind=ConfigMap` or `kind=Secret`)
+4. AS a KES user i want to fetch **from multiple** stores and store the secrets **in a single** Frontend
+5. AS a KES operator i want to limit the access to certain stores or subresources (e.g. having one central KES instance that handles all ES - similar to `iam.amazonaws.com/permitted` annotation per namespace)
+
+### Stores
+
+These stores are relevant:
+* AWS Secure Systems Manager Parameter Store
+* AWS Secrets Manager
+* Hashicorp Vault
+* Azure Key Vault
+* Alibaba Cloud KMS Secret Manager
+* Google Cloud Platform Secret Manager
+* Kubernetes (see #422)
+* noop (see #476)
+
+### Frontends
```

Comment:
Are the file frontends (https://github.com/godaddy/kubernetes-external-secrets/issues/81) interesting to bring up here? I guess thats more of a `noop` frontend tho. So you can have a side-car fetch populate a volume with the secret data without it ending up in a k8s secret.

Reply:
I feel like this is out of scope for now (this is a really big feature)

Reply:
(this was added to the backlog)